gnomad/utils/transcript_annotation.py

```diff
@@ -361,6 +433,18 @@ def tx_aggregate_variants(
     if additional_group_by is not None:
         grouping = grouping + list(additional_group_by)
 
+    # Previous steps of the transcript annotation pipeline require that the input ht is
+    # keyed by locus and alleles so that the correct transcripts are retained when
+    # filtering variants by transcript (with `tx_filter_variants_by_csqs`) and exploding
+    # the VEP annotation (in `tx_annotate_variants`). However, if "alleles" is not
+    # present in the additional_group_by, a transcript can be counted multiple times if
+    # it is associated with multiple alleles since the transcript expression is keyed
+    # by transcript_id and gene_id. To avoid this, we re-key the tx_ht by locus,
+    # gene_id, transcript_id, and any additional_group_by fields, followed by a distinct
+    # operation to ensure that each transcript is only counted once per locus.
+    if "alleles" not in additional_group_by:
+        ht = ht.key_by(*grouping, "transcript_id").distinct()
```

Comment:
sorry for the confusion -- I was initially trying to ask whether the values at each loci for each transcript were the same.

I'm a bit confused now; I think the previous method (of keying by just locus and transcript) deduplicates, this new method could count a transcript multiple times per locus. For example, if `additional_group_by = "lof"`, then grouping at this point would be `["locus", "gene_id", "lof"]`. That means HT would be keyed by `locus, gene_id, lof, transcript_id` here, which means a transcript could still get counted multiple times if one variant in the transcript is pLoF and another variant isn't, right? let me know if I've missed something

Reply:
> sorry for the confusion -- I was initially trying to ask whether the values at each loci for each transcript were the same.

Yeah, I figured that's what you were asking, and the values for the expression proportions will be the same, but other columns can be different, so by doing the distinct I could be removing some of the values that need to be counted. It's not a problem for the main case where we use the function, but if people change up the additional_grouping, it could return the wrong result.

Here is an example of when the previous method would have failed:

`tx_ht`:
<img width="255" alt="Screenshot 2024-08-14 at 9 53 43 AM" src="https://github.com/user-attachments/assets/38b37c6a-b295-4db5-ac41-b896f7627d86">

`context_ht` (after exploding by transcript):
<img width="350" alt="Screenshot 2024-08-14 at 9 53 47 AM" src="https://github.com/user-attachments/assets/4992a8c8-37e1-4201-a0b6-8dad4873018a">

annotate `context_ht` with `tx_ht`:
<img width="454" alt="Screenshot 2024-08-14 at 9 53 52 AM" src="https://github.com/user-attachments/assets/340ea5e3-79c4-4644-8bd5-a70ae26b8c73">

Then if we do the distinct it in the way we would for the base level, it's basically this:
```
always_grouping = ["locus", "gene_id"]
additional_grouping = []
grouping = always_grouping + additional_grouping

context_ht = context_ht.key_by("locus", "transcript_id").distinct()
```
<img width="454" alt="Screenshot 2024-08-14 at 9 53 57 AM" src="https://github.com/user-attachments/assets/34e78779-21a3-47ac-ab38-72f57e791c7d">

And then the aggregation:
```
agg_ht = context_ht.group_by(*grouping).aggregate(
    sum_exp_prop_mean=hl.agg.sum(context_ht.exp_prop_mean)
)
```
<img width="304" alt="Screenshot 2024-08-14 at 9 54 02 AM" src="https://github.com/user-attachments/assets/f6e5e74d-44a3-410d-b714-f5206f6854e3">

That is fine because, it doesn't matter which `lof` we grab.



However, if we change the `additional_grouping` to include "lof":
```
always_grouping = ["locus", "gene_id"]
additional_grouping = ["lof"]
grouping = always_grouping + additional_grouping

context_ht = context_ht.key_by("locus", "transcript_id").distinct()
```
<img width="449" alt="Screenshot 2024-08-14 at 10 00 27 AM" src="https://github.com/user-attachments/assets/d3b3bbda-3e01-4696-a208-c00443eadd70">

The aggregation gives:
```
agg_ht = context_ht.group_by(*grouping).aggregate(
    sum_exp_prop_mean=hl.agg.sum(context_ht.exp_prop_mean)
)
```
<img width="334" alt="Screenshot 2024-08-14 at 10 00 31 AM" src="https://github.com/user-attachments/assets/7ae9b323-9265-4031-a4d4-32328045771e">

What we want is: 
```
context_ht = context_ht.key_by(*grouping, "transcript_id").distinct()
```
<img width="460" alt="Screenshot 2024-08-14 at 10 02 16 AM" src="https://github.com/user-attachments/assets/9ad0ae75-d20f-436f-999f-a05a2a119928">

So the aggregation gives: 
<img width="345" alt="Screenshot 2024-08-14 at 10 02 20 AM" src="https://github.com/user-attachments/assets/097d54fb-a7fa-47a9-8750-3455f0ac6b3e">

Reply:
thank you for this! it sounds like the concern isn't about accidentally duplicating transcript representation at a locus but more about making sure the values are represented as requested by the user with `additional_groupings` -- maybe this comment:
```
    # Previous steps of the transcript annotation pipeline require that the input ht is
    # keyed by locus and alleles so that the correct transcripts are retained when
    # filtering variants by transcript (with `tx_filter_variants_by_csqs`) and exploding
    # the VEP annotation (in `tx_annotate_variants`). However, if "alleles" is not
    # present in the additional_group_by, a transcript can be counted multiple times if
    # it is associated with multiple alleles since the transcript expression is keyed
    # by transcript_id and gene_id. To avoid this, we re-key the tx_ht by locus,
    # gene_id, transcript_id, and any additional_group_by fields, followed by a distinct
    # operation to ensure that each transcript is only counted once per locus.
```
could be edited to be:
```
    # Previous steps of the transcript annotation pipeline require that the input ht is
    # keyed by locus and alleles so that the correct transcripts are retained when
    # filtering variants by transcript (with `tx_filter_variants_by_csqs`) and exploding
    # the VEP annotation (in `tx_annotate_variants`). However, if "alleles" is not
    # present in the additional_group_by, a transcript that is associated with multiple alleles
    # at a locus may not be represented correctly after deduplicating locus and transcript
    # combinations with a distinct operation, since this operation deduplicates by selecting
    # a random row. To ensure the desired values are selected during `distinct()` filter, 
    # we re-key the tx_ht by locus, gene_id, transcript_id, and any additional_group_by fields
    # prior to running `distinct()'.
```
or another rewording?